### PR TITLE
"project-report-implementation-script"

### PR DESCRIPTION
**Pull request by MallanagoudaB**
# Description
These recommendations are intended to promote code quality and team communication during software development. They cover a variety of topics, including ensuring that pull requests are submitted to the correct branch, documenting new methods, preserving consistency across many services, and avoiding typical blunders like accessing APIs or DB queries within loops. Sensitive data should not be uploaded, and changes to environment variables or database models should be executed consistently. Teams may work more effectively and develop higher-quality software by adhering to these standards.

## Type of change
Please choose appropriate options.

- [ ]  Bug fix (non-breaking change which fixes an issue)
- [ ]  New feature (non-breaking change which adds functionality)
- [ ]  Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ]  Enhancement (additive changes to improve performance)
- [ ]  This change requires a documentation update

## Checklist 

- [ ]  It's critical to avoid making needless file modifications in contributions, such as adding new lines, console logs, or additional spaces, to guarantee cleaner and more efficient code. Furthermore, eliminating unnecessary imports from a file might enhance code readability and efficiency.
- [ ]  Ensure that the pull request is assigned to the right base branch and that the development branch name contains the JIRA Task Id. Furthermore, each commit message should include the JIRA Task Id in the manner "ED-100: message".
- [ ]  Only update packages if it is mentioned and authorized in the design document, and make sure that you have the required permissions.
- [ ]  Avoid making API and database queries inside a loop as it can lead to performance issues and slow down the system.
- [ ]  When calling another function inside a given function, add comments explaining the purpose and meaning of the passed arguments and expected return values.
- [ ]  If adding a blank argument in a function, add a comment explaining the reason for the blank argument.
- [ ]  Before submitting a pull request, do a self-review of your code to ensure there are no conflicts with the base branch and all comments have been addressed.
- [ ]  Before merging a pull request, it's important to have other team members review it to catch any potential errors or issues
- [ ]  To maintain code integrity, it's important to remove all related changes when removing code during a code review.
- [ ]  If new constants, endpoints, or utility functions are introduced, it is important to check if they already exist in the service to avoid any duplication.
- [ ]  Whenever a new environment variable is added to a service, it's important to ensure that the necessary changes are made to related files such as ".env.sample" and "envVariables.js" to maintain consistency and avoid errors. Additionally, the new environment variable should be added to the devops repository to ensure that it is properly documented and accessible to the team.
- [ ]  When adding a new function to a service, it is important to document it with relevant information such as the name, parameters, and return value in a consistent format across all services. Additionally, if there are any changes to the API response, ensure that the documentation in the controllers is updated accordingly. 
- [ ]  Write a clear and concise commit message that describes the changes made.
- [ ]  Maintain consistent function signature and code across all services when adding a function to multiple services. Implement changes to database models in all services that use the same model. 
- [ ]  Use only let and const. Do not use var.
- [ ]  Make common functions for repetitive code blocks.
- [ ]  Avoid uploading sensitive information such as secret tokens or passwords in pull requests to ensure data security.
- [ ]  Maintain consistent indentation and spacing throughout the code.

### CODE REVIEW

File: Project-Service implementation-Script/common_config/config.ini [hunk 1]
```diff
@@ -0,0 +1,133 @@
+[pre-prod]
+host = http://localhost:3569/
```
Comment (adithyadinesh0412) [inline Project-Service implementation-Script/common_config/config.ini:2]: Should be removed

Comment (adithyadinesh0412) [reply Project-Service implementation-Script/common_config/config.ini:2]: All this was discussed before

Comment (adithyadinesh0412) [inline Project-Service implementation-Script/common_config/config.ini:2]: I think this folder is not required. 
@MallanagoudaB

File: Project-Service-implementation-Script/main.py [hunk 2]
```diff
@@ -0,0 +1,5242 @@
...
+        print(e)
+        return False
+
+# Generate access token for the APIs. 
+def generateAccessToken(solutionName_for_folder_path):
+    # production search user api - start
+    headerKeyClockUser = {'Content-Type': config.get(environment, 'content-type')}
+    # responseKeyClockUser = requests.post(url=config.get(environment, 'elevateuserhost') + config.get(environment, 'userlogin'), headers=headerKeyClockUser,
+                                        #  data=json.dumps(config.get(environment, 'keyclockAPIBody')))
+    # terminatingMessage(type(json.loads(config.get(environment, 'keyclockAPIBody'))))\
```
Comment (adithyadinesh0412) [inline Project-Service-implementation-Script/main.py:551]: remove comments if not required
@MallanagoudaB

File: Project-Service-implementation-Script/main.py [hunk 3]
```diff
@@ -0,0 +1,5242 @@
...
+        fileheader = ["Access Token","Access Token succesfully genarated","Passed"]
+        apicheckslog(solutionName_for_folder_path,fileheader)
+        print("--->Access Token Generated!")
+        return accessTokenUser
+    
+    print("Error in generating Access token")
+    print("Status code : " + str(responseKeyClockUser.status_code))
+    createAPILog(solutionName_for_folder_path, messageArr)
+    fileheader = ["Access Token", "Error in generating Access token", "Failed", str(responseKeyClockUser.status_code) + " Check access token api"]
+    # fileheader = ["Access Token", "Error in generating Access token", "Failed",responseKeyClockUser.status_code+"Check access token api"]
```
Comment (adithyadinesh0412) [inline Project-Service-implementation-Script/main.py:575]: Comment is required?

**Review by adithyadinesh0412 — COMMENTED** [review]
Can approve only after testing the code with the requirement.

File: Project-Service implementation-Script/common_config/config.ini (hunk 1 above)
Comment (MallanagoudaB) [reply Project-Service implementation-Script/common_config/config.ini:2]: @adithyadinesh0412  sorry but its a old file the new file is config.sample please check

File: Project-Service-implementation-Script/main.py [hunk 4]
```diff
@@ -0,0 +1,5242 @@
...
+    apicheckslog(solutionName_for_folder_path, fileheader)
+    fileheader = ["Access Token", "Error in generating Access token", "Failed","Check Headers of api"]
+    apicheckslog(solutionName_for_folder_path, fileheader)
+    terminatingMessage("Please check API logs.")
+
+# Function to search for programs
+def getProgramInfo(accessTokenUser, solutionName_for_folder_path, programNameInp):
+    global programID, programExternalId, programDescription, isProgramnamePresent, programName
+    programName = programNameInp
+    programUrl = config.get(environment, 'elevateprojecthost') + config.get(environment, 'fetchProgramInfoApiUrl') + programNameInp.lstrip().rstrip()
```
[the remark this answers is not in the document]
Comment (VISHNUDAS-tunerlabs) [reply Project-Service-implementation-Script/main.py:581]: @adithyadinesh0412 we don't have any dependency over ml-core-service. program related apis are included in project-service.

File: Project-Service-implementation-Script/main.py [hunk 5]
```diff
@@ -0,0 +1,5238 @@
...
+        API_log.close()
+
+    API_log = open(file_exists, "a", encoding='utf-8')
+    API_log.write("\n")
+    for msg in messageArr:
+        API_log.write(msg)
+        API_log.write("\n")
+    API_log.close()
+
+def apicheckslog(solutionName_for_folder_path, messageArr):
```
Comment (VISHNUDAS-tunerlabs) [inline Project-Service-implementation-Script/main.py:677]: @MallanagoudaB Please add function definition

File: Project-Service-implementation-Script/main.py [hunk 6]
```diff
@@ -0,0 +1,5238 @@
...
+            print(roledetails)
+        else:
+            terminatingMessage("-->Given username/email is not present in DIKSHA platform<--.")
+    else:
+        terminatingMessage("User fetch API failed. Check logs.")
+    return [userKeycloak, userName, firstName,roledetails,rootOrgName,rootOrgId]
+
+
+# fetch org Ids 
+def fetchOrgId(environment, accessToken, parentFolder, OrgName):
```
Comment (VISHNUDAS-tunerlabs) [inline Project-Service-implementation-Script/main.py:732]: @MallanagoudaB is this function used any where. If yes can you please point out the line number where it is used

File: Project-Service-implementation-Script/main.py [hunk 7]
```diff
@@ -0,0 +1,5238 @@
...
+    return orgIds
+
+
+# Print message and terminate the program
+def terminatingMessage(msg):
+    print(msg)
+    sys.exit()
+
+
+def fetchEntityId(solutionName_for_folder_path, accessToken, entitiesNameList, scopeEntityType):
```
Comment (VISHNUDAS-tunerlabs) [inline Project-Service-implementation-Script/main.py:785]: @MallanagoudaB function description is missing

File: Project-Service-implementation-Script/common_config/config.ini [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+criteriarubricuploadapiurl = private/mlsurvey/api/v1/solutions/uploadCriteriaRubricExpressions/
+themerubricuploadapiurl = private/mlsurvey/api/v1/solutions/uploadThemesRubricExpressions/
+importsurveysolutiontemplateurl = private/mlsurvey/api/v1/surveys/importSurveryTemplateToSolution/
+importsurveysolutiontoprogramurl = private/mlsurvey/api/v1/surveys/mapSurverySolutionToProgram/
+solutiontoprogrammappingapiurl = private/mlsurvey/api/v1/solutions/importFromSolution
+fetchprograminfoapiurl = project/v1/programs/list?page=1&limit=5&search=
+fetchsolutiondetails = project/v1/solutions/list?page=1&limit=100&search=&type=observation&subType
+frameworkcreationapi = private/mlsurvey/api/v1/frameworks/create
+solutionupdateapi = project/v1/solutions/update/
+listofrolesapi = private/mlsurvey/api/v1/userRoles/list
```
Comment (VISHNUDAS-tunerlabs) [inline Project-Service-implementation-Script/common_config/config.ini:107]: @MallanagoudaB is this endpoint used?

File: Project-Service-implementation-Script/common_config/config.ini [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+taskuploadapi = project/v1/project/templateTasks/bulkCreate/
+projectsolutioncreationapi = project/v1/solutions/create
+mapsolutiontoproject = project/v1/project/templates/importProjectTemplate/
+fetchsolutiondoc = project/v1/solutions/getDetails/
+programcreationurl = project/v1/programs/create
+pdpmurl = private/mlsurvey/api/v1/userExtension/bulkUpload
+fetchlink = project/v1/solutions/fetchLink/
+courseprogrammapping = private/mlcore/api/v1/solutions/create
+readcourseurl = api/content/v1/read/
+fetchsolutiondump = private/mlcore/api/v1/solutions/getDetails/
```
Comment (VISHNUDAS-tunerlabs) [inline Project-Service-implementation-Script/common_config/config.ini:120]: @MallanagoudaB is this api used?

File: Project-Service-implementation-Script/common_config/config.ini [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+fetchorgdetails = api/org/v1/search
+fetchprojectlist = private/mlprojects/api/v1/library/categories/projects
```
Comment (VISHNUDAS-tunerlabs) [inline Project-Service-implementation-Script/common_config/config.ini:122]: @MallanagoudaB why we are still having mlprojects end point?

File: Project-Service-implementation-Script/main.py [hunk 11]
```diff
@@ -0,0 +1,5238 @@
...
+
+
+def themesUpload(solutionName_for_folder_path, wbObservation, millisAddObs, accessToken, frameworkExternalId,obsWORubWS):
+    global dictCritLookUp
+    with open(solutionName_for_folder_path + '/criteriaUpload/uploadInternalIdsSheet.csv', 'r',encoding='utf-8') as criteriaInternalFile:
+        criteriaInternalReader = csv.DictReader(criteriaInternalFile)
+        for crit in criteriaInternalReader:
+            dictCritLookUp[crit['Criteria External Id']] = crit['Criteria Internal Id']
+    if obsWORubWS:
+        print("Themes Observation without rubrics with scores")
```
Comment (VISHNUDAS-tunerlabs) [inline Project-Service-implementation-Script/main.py:1759]: @MallanagoudaB why we are having observation related code in implementation script? why is it required?

File: Project-Service-implementation-Script/main.py [hunk 12]
```diff
@@ -0,0 +1,5238 @@
...
+        xfile.save(MainFilePath + "/" + programFile + '-SuccessSheet.xlsx')
+        print("Program success sheet is created")
+
+    else:
+        print("Fetch solution link API Failed")
+        messageArr.append("Response : " + str(responseFetchSolutionLinkApi.text))
+        createAPILog(solutionName_for_folder_path, messageArr)
+        sys.exit()
+
+def prepareSuccessSheet(solutionName_for_folder_path, filePathAddObs, observationExternalId, millisAddObs):
```
Comment (VISHNUDAS-tunerlabs) [inline Project-Service-implementation-Script/main.py:3058]: @MallanagoudaB function definition please

File: Project-Service-implementation-Script/main.py [hunk 13]
```diff
@@ -0,0 +1,5238 @@
...
+                            solutionId = None
+                            solutionId = responseCreateSolutionApi["result"]["solutionId"]
+                            bodySolutionUpdate = {"creator": dictDetailsEnv['Name_of_the_creator'].encode('utf-8').decode('utf-8')}
+                            solutionUpdate(parentFolder, accessToken, solutionId, bodySolutionUpdate)
+                            return [solutionId, surveySolutionExternalId]
+                        else:
+                            terminatingMessage("Survey creation Failed, check logs!")
+
+# upload survey questions 
+def uploadSurveyQuestions(parentFolder, wbSurvey, addObservationSolution, accessToken, surveySolutionExternalId, surveyParentSolutionId,millisecond):
```
Comment (VISHNUDAS-tunerlabs) [inline Project-Service-implementation-Script/main.py:3310]: @MallanagoudaB We clearly discussed that we have to remove observation and survey related code if it is not required. this script only for project-service. why this is not taken care?

**Review by VISHNUDAS-tunerlabs — CHANGES_REQUESTED** [review]
3rd review on June 12

File: Project-Service-implementation-Script/common_config/config.ini [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+editsvgtemp = private/mlcore/api/v1/certificateTemplates/createSvg?baseTemplateId=
+uploadcertificatetosvg = private/mlcore/api/v1/certificateTemplates/uploadTemplate/
+updatecertificatesolu = private/mlcore/api/v1/solutions/update/
+updateprojecttemplate = private/mlprojects/api/v1/project/templates/update/
+
+onelogo-onesignature = 1L1S
+onelogo-twosignature = 1L2S
+twologo-onesignature = 2L1S
+twologo-twosignature = 2L2S
+[local]
```
[the remark this answers is not in the document]
Comment (MallanagoudaB) [reply Project-Service-implementation-Script/common_config/config.ini:64]: @VISHNUDAS-tunerlabs  please refer config.sample file for this

File: Project-Service-implementation-Script/common_config/config.sample [hunk 15]
```diff
@@ -0,0 +1,72 @@
+[dev]
```
[the remark this answers is not in the document]
Comment (MallanagoudaB) [reply Project-Service-implementation-Script/common_config/config.sample:1]: OKay i will add

File: Project-Service-implementation-Script/main.py [hunk 16]
```diff
@@ -0,0 +1,5238 @@
...
+            for row_index_env in range(2, detailsEnvSheet.nrows):
+                dictDetailsEnv = {keysEnv[col_index_env]: detailsEnvSheet.cell(row_index_env, col_index_env).value
+                                  for
+                                  col_index_env in range(detailsEnvSheet.ncols)}
+                programNameInp = dictDetailsEnv['Title of the Program'].encode('utf-8').decode('utf-8') if dictDetailsEnv['Title of the Program'] else terminatingMessage("\"Title of the Program\" must not be Empty in \"Program details\" sheet")
+
+            extIdPGM = dictDetailsEnv['Program ID'].encode('utf-8').decode('utf-8') if dictDetailsEnv['Program ID'] else terminatingMessage("\"Program ID\" must not be Empty in \"Program details\" sheet")
+
+            programdesigner = dictDetailsEnv['Diksha username/user id/email id/phone no. of Program Designer'].encode('utf-8').decode('utf-8') if dictDetailsEnv['Program ID'] else terminatingMessage("\"Diksha username/user id/email id/phone no. of Program Designer\" must not be Empty in \"Program details\" sheet")
+            # userDetails = fetchUserDetails(environment, accessToken, programdesigner)
```
[the remark this answers is not in the document]
Comment (MallanagoudaB) [reply Project-Service-implementation-Script/main.py:216]: as i mentioned you before user api was not working so you said me to skip .

File: Project-Service-implementation-Script/common_config/config.ini (hunk 8 above)
Comment (MallanagoudaB) [reply Project-Service-implementation-Script/common_config/config.ini:107]: please refer config.sample

File: Project-Service-implementation-Script/common_config/config.ini (hunk 9 above)
Comment (MallanagoudaB) [reply Project-Service-implementation-Script/common_config/config.ini:120]: please refer config.sample

File: Project-Service-implementation-Script/common_config/config.ini (hunk 10 above)
Comment (MallanagoudaB) [reply Project-Service-implementation-Script/common_config/config.ini:122]: please refer config.sample

File: Project-Service-implementation-Script/main.py [hunk 17]
```diff
@@ -0,0 +1,5238 @@
...
+                       range(detailsEnvSheet.ncols)]
+            for row_index_env in range(2, detailsEnvSheet.nrows):
+                dictDetailsEnv = {keysEnv[col_index_env]: detailsEnvSheet.cell(row_index_env, col_index_env).value
+                                  for
+                                  col_index_env in range(detailsEnvSheet.ncols)}
+                programNameInp = dictDetailsEnv['Title of the Program'].encode('utf-8').decode('utf-8') if dictDetailsEnv['Title of the Program'] else terminatingMessage("\"Title of the Program\" must not be Empty in \"Program details\" sheet")
+
+            extIdPGM = dictDetailsEnv['Program ID'].encode('utf-8').decode('utf-8') if dictDetailsEnv['Program ID'] else terminatingMessage("\"Program ID\" must not be Empty in \"Program details\" sheet")
+
+            programdesigner = dictDetailsEnv['Diksha username/user id/email id/phone no. of Program Designer'].encode('utf-8').decode('utf-8') if dictDetailsEnv['Program ID'] else terminatingMessage("\"Diksha username/user id/email id/phone no. of Program Designer\" must not be Empty in \"Program details\" sheet")
```
[the remark this answers is not in the document]
Comment (MallanagoudaB) [reply Project-Service-implementation-Script/main.py:215]: in csv template its a old template if you provide me new updated csv then i can make changes to elevateProject

File: Project-Service-implementation-Script/main.py [hunk 18]
```diff
@@ -0,0 +1,5238 @@
...
+            detailsEnvSheet = wbPgm.sheet_by_name(sheetEnv)
+            keysEnv = [detailsEnvSheet.cell(1, col_index_env).value for col_index_env in
+                       range(detailsEnvSheet.ncols)]
+            for row_index_env in range(2, detailsEnvSheet.nrows):
+                dictDetailsEnv = {keysEnv[col_index_env]: detailsEnvSheet.cell(row_index_env, col_index_env).value
+                                  for
+                                  col_index_env in range(detailsEnvSheet.ncols)}
+
+                if str(dictDetailsEnv['Is a SSO user?']).strip() == "YES":
+                    programmanagername2 = dictDetailsEnv['Diksha user id ( profile ID)'] if dictDetailsEnv['Diksha user id ( profile ID)'] else terminatingMessage("\"Diksha user id ( profile ID)\" must not be Empty in \"Program details\" sheet")
```
[the remark this answers is not in the document]
Comment (MallanagoudaB) [reply Project-Service-implementation-Script/main.py:245]: in csv template its a old template if you provide me new updated csv then i can make changes to elevateProject

[the remark this answers is not in the document]
Comment (MallanagoudaB) [reply Project-Service-implementation-Script/main.py:245]: in csv template its a old template if you provide me new updated csv then i can make changes to elevateProject

File: Project-Service-implementation-Script/main.py [hunk 19]
```diff
@@ -0,0 +1,5238 @@
...
+                else:
+                    try :
+                        programmanagername2 = dictDetailsEnv['Login ID on DIKSHA'].encode('utf-8').decode('utf-8') if dictDetailsEnv['Login ID on DIKSHA'] else terminatingMessage("\"Login ID on DIKSHA\" must not be Empty in \"Program details\" sheet")
```
[the remark this answers is not in the document]
Comment (MallanagoudaB) [reply Project-Service-implementation-Script/main.py:248]: in csv template its a old template if you provide me new updated csv then i can make changes to elevateProject

File: Project-Service-implementation-Script/README.md [hunk 20]
```diff
@@ -0,0 +1,31 @@
+# Python scripts for elevate-project implementation
+
+Python script to upload a program and add multiple resources like Projects , Surveys , Observations (with and without rubrics) to it and also upload project add multiple resources like task.
```
Comment (VISHNUDAS-tunerlabs) [inline Project-Service-implementation-Script/README.md:3]: @MallanagoudaB we only have project and tasks here

File: Project-Service-implementation-Script/README.md [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+
+### Resource templates
+- [Programs](https://docs.google.com/spreadsheets/d/1_pEomGIWxVZYwneQwY606zt26C72p9DU/edit?usp=drive_link&ouid=113799545932705393937&rtpof=true&sd=true)
```
Comment (VISHNUDAS-tunerlabs) [inline Project-Service-implementation-Script/README.md:6]: @MallanagoudaB Please add the files with proper filename, directory as required

File: Project-Service-implementation-Script/main.py [hunk 22]
```diff
@@ -0,0 +1,4668 @@
+import imaplib
```
Comment (VISHNUDAS-tunerlabs) [inline Project-Service-implementation-Script/main.py:1]: @MallanagoudaB Please verify the user validation for program manager or program designer is happening

**Review by VISHNUDAS-tunerlabs — CHANGES_REQUESTED** [review]
reviewed on june 14 (12:8) TTF : 20 RPT: 5

File: Project-Service-implementation-Script/README.md [hunk 23]
```diff
@@ -0,0 +1,41 @@
...
+``` source { relative path to virtualEnv}/bin/activate ```
+In Windows
+``` { relative path to virtualEnv}/Scripts/activate ```
+- Install all the dependencies using requirement.txt using following command. 
+```  pip3 install -r requirement.txt ```
+- Make sure there are no errors in the install.
+- If there are any errors in the install, try to install the same version of the libraries seperatly.
+- Download the user given template and save it in the same file where the code is hosted.
+- There are TWO Command to run the script.
+i. For programTemplate :
```
Comment (VISHNUDAS-tunerlabs) [inline Project-Service-implementation-Script/README.md:27]: @MallanagoudaB Please add explanation on what all commands will do

File: Project-Service-implementation-Script/README.md [hunk 24]
```diff
@@ -0,0 +1,41 @@
+# Python scripts for elevate-project implementation
```
Comment (VISHNUDAS-tunerlabs) [inline Project-Service-implementation-Script/README.md:1]: @MallanagoudaB change MEW.xlsx file name to some relevant name, Please

**Review by VISHNUDAS-tunerlabs — CHANGES_REQUESTED** [review]

**Pull request merged**